11-type-constructors/algo/src/External.hs

```diff
@@ -0,0 +1,99 @@
+{-# LANGUAGE TypeFamilies #-}
+
+module External where
+
+import Common
+import Control.Exception.Base
+
+-- The external expression langauge
+data Exp
+  = EVar TID
+  | Eλ TID Typ Exp
+  | EAp Exp Exp
+  | ETypLet TID Typ Exp
+  | EExpLet TID Typ Exp Exp
+  deriving (Show)
+
+-- The external type language
+data Typ
+  = TVar TID
+  | Bse -- We'll remove this once built in types work
+  | Typ :⊕ Typ
+  | ETHole HID
+  | NETHole HID Typ
+  | Tλ TID Knd Typ
+  | TAp Typ Typ
+  deriving (Show)
+
+data Knd
+  -- the external kind language
+  = Type
+  | KHole
+  | S Knd Typ
+  | Π TID Knd Knd
+  deriving (Show)
+
+-- I'm afraid that sooner or later I'll run into a case where just using De
```

Comment:
why didn't you use de Bruijn indices? (that's what we're doing in the Hazel implementation)

Reply:
At the time, I thought:
-- the phi uses variable names, so the implementation might as well too to make sure the renaming rules are right
-- it might be easier to debug w/o de Bruijn indices
-- I was going to use a Barendregt convention anyways, so it should be a little easier to implement
(of course the third point doesn't actually make sense, and is somewhat opposed to the first)

I agree in retrospect that I should've done de Bruijn indices like you'd said at the beginning, but I was still learning these things and wrote all the renaming stuff early on

Now there's not much point in changing it in the haskell implementation right?